File: 0100-same-tree/0100-same-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # Time complexity : O(n) not balanced and O(logn) balanced 
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        # if both tree empty:
        #   return True
        if not p and not q:
            return True
        # if one of the tree is empty:
        #   return False
        if not p or not q:
            return False
        
        # if value of p not equal value of q:
        #   return False
        if p.val != q.val:
            return False
        
        # return ( call recursive on p.left and q.left AND
        # call recursive on p.right and q.right)
        return (self.isSameTree(p.left, q.left) and
               self.isSameTree(p.right,q.right))
        
        
        

File: 0100-same-tree/0100-same-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isSameTree(p: TreeNode | null, q: TreeNode | null): boolean {
    // BASE CASE: if both tree are empty return true
    if (!p && !q) {
        return true;
    }
    // if one of the tree is empty return false
    if (!p || !q) {
        return false;
    }
    // if val at first tree not equal val of second tree return false
    if (p.val !== q.val) {
        return false;
    }
    // recurse the (left of first tree,left of second tree) AND
    //             (right of first tree, right of second tree) 
    return isSameTree(p.left,q.left) && isSameTree(p.right,q.right);
};